server/src/main/scala/org/http4s/server/middleware/VirtualHost.scala

```diff
@@ -0,0 +1,41 @@
+package org.http4s
+package server
+package middleware
+
+import headers.Host
+import Status.BadRequest
+
+/** Middleware for virtual host mapping
+  *
+  * The `VirtualHost` middleware allows multiple services to be mapped
+  * based on the [[Host]] header of the [[Request]]. A default service
+  * is specified which will handle cases where the [[Host]] header is
+  * not present (should only happen for HTTP/1.0 requests) or if none
+  * of the hosts match.
+  * If the port is not specified by the client _or_ the server, it is
+  * ignored because the middleware cannot know the default port for the
+  * underlying transport.
+  */
+object VirtualHost {
+
+  /** Specification of the virtual host */
+  case class HostService(service: HttpService, host: String, port: Option[Int] = None)
+
+  def apply(default: HostService, rest: HostService*): HttpService = {
+
+    def check(host: Host, hs: String, port: Option[Int]): Boolean =
+      host.host == hs && (port.isEmpty || host.port.isEmpty || host.port.get == port.get)
+
+    val all = (default +: rest).toVector
+
+    Service.lift { req =>
+      req.headers.get(Host) match {
```

Comment:
[RFC 7230: 5.4](https://tools.ietf.org/html/rfc7230#section-5.4) says:
> A client __MUST__ send a Host header field in all HTTP/1.1 request
> messages.  If the target URI includes an authority component, then a
> client __MUST__ send a field-value for Host that is identical to that
> authority component, excluding any userinfo subcomponent and its "@"
> delimiter (Section 2.7.1)."

Should we be more strict on requiring the URI and Host header agree? 
I do think it would be proper to look for the host information from a HTTP/1.0 with an absolute URI.